library/bignum_core.c

```diff
@@ -291,4 +293,139 @@ int mbedtls_mpi_core_write_be( const mbedtls_mpi_uint *X,
     return( 0 );
 }
 
+void mbedtls_mpi_core_montmul( mbedtls_mpi_uint *X,
+                               const mbedtls_mpi_uint *A,
+                               const mbedtls_mpi_uint *B,
+                               size_t B_limbs,
+                               const mbedtls_mpi_uint *N,
+                               size_t AN_limbs,
+                               mbedtls_mpi_uint mm,
+                               mbedtls_mpi_uint *T )
+{
+    memset( T, 0, ( 2 * AN_limbs + 1 ) * ciL );
+
+    for( size_t i = 0; i < AN_limbs; i++, T++ )
+    {
+        mbedtls_mpi_uint u0, u1;
+        /* T = (T + u0*B + u1*N) / 2^biL */
+        u0 = A[i];
+        u1 = ( T[0] + u0 * B[0] ) * mm;
+
+        (void) mbedtls_mpi_core_mla( T, AN_limbs + 2, B, B_limbs, u0 );
+        (void) mbedtls_mpi_core_mla( T, AN_limbs + 2, N, AN_limbs, u1 );
+    }
+
+    /* It's possible that the result in T is > N, and so we might need to subtract N */
+
+    mbedtls_mpi_uint carry  = T[AN_limbs];
```

Comment:
I disagree. There are three lines of code, one of which doesn't really do much (the assignment to carry). It's all very specific to the Montgomery multiplication, and in fact, of the 4 possible values we could have for (carry, borrow), one pair is not possible due to the code before. Sure, we can separate into an auxiliary function if we find other places we do exactly this, but I don't see anywhere else it could be reused for now.

Reply:
Having an auxiliary function would be primarily about understanding what's going on, and only secondarily about reusability. But now with a better explanation I think it's ok.